python/cudf_polars/cudf_polars/dsl/nodebase.py

```diff
@@ -0,0 +1,148 @@
+
+    from typing_extensions import Self
+
+
+__all__: list[str] = ["Node"]
+
+
+class Node:
+    """
+    An abstract node type.
+
+    Nodes are immutable!
+
+    This contains a (potentially empty) tuple of child nodes,
+    along with non-child data. For uniform reconstruction and
+    implementation of hashing and equality schemes, child classes need
+    to provide a certain amount of metadata when they are defined.
+    Specifically, the ``_non_child`` attribute must list, in-order,
+    the names of the slots that are passed to the constructor. The
+    constructor must take arguments in the order ``(*_non_child,
+    *children).``
+    """
+
+    __slots__ = ("_hash_value", "_repr_value", "children")
+    _hash_value: int
+    _repr_value: str
+    children: tuple[Node, ...]
+    _non_child: ClassVar[tuple[str, ...]] = ()
+
+    def _ctor_arguments(self, children: Sequence[Node]) -> Sequence:
+        return (*(getattr(self, attr) for attr in self._non_child), *children)
+
+    def reconstruct(
+        self, children: Sequence[Node]
+    ) -> Self:  # pragma: no cover; not yet used
+        """
+        Rebuild this node with new children.
+
+        Parameters
+        ----------
+        children
+            New children
+
+        Returns
+        -------
+        New node with new children. Non-child data is shared with the input.
+        """
+        return type(self)(*self._ctor_arguments(children))
+
+    def get_hashable(self) -> Hashable:
+        """
+        Return a hashable object for the node.
+
+        Returns
+        -------
+        Hashable object.
+
+        Notes
+        -----
+        This method is used by the :meth:`__hash__` implementation
+        (which does caching). If your node type needs special-case
+        handling for some of its attributes, override this method, not
+        :meth:`__hash__`.
+        """
+        return (type(self), self._ctor_arguments(self.children))
+
+    def __hash__(self) -> int:
+        """
+        Hash of an expression with caching.
+
+        See Also
+        --------
+        get_hashable
+        """
+        try:
+            return self._hash_value
+        except AttributeError:
+            self._hash_value = hash(self.get_hashable())
+            return self._hash_value
+
+    def is_equal(self, other: Any) -> bool:
+        """
+        Equality of two expressions.
+
+        Override this in subclasses, rather than :meth:`__eq__`.
+
+        Parameter
+        ---------
+        other
+            object to compare to.
+
+        Notes
+        -----
+        Since nodes are immutable, this does common subexpression
+        elimination when two nodes are determined to be equal.
+
+        Returns
+        -------
+        True if the two expressions are equal, false otherwise.
+        """
+        if self is other:
+            return True
+        if type(self) is not type(other):
+            return False  # pragma: no cover; __eq__ trips first
```

Comment:
Generally no. I guess I can tighten the type of `other`.

Reply:
Oh, is this branch here just to narrow the type so that mypy doesn't complain on subsequent lines? If so, yes it seems like you ought to be able to simply define other as a more specific type in the signature unless you intend for this method to also be safe to use outside of `__eq__` , in which case leaving it as is seems fine.

Reply:
No, I'll add a docstring note and remove the type narrowing.